include/thread_pool/thread_pool.h

```diff
@@ -82,8 +84,9 @@ namespace dp {
                             priority_queue_.rotate_to_front(id);
                             // check if all tasks are completed and release the barrier (binary
                             // semaphore)
-                            if (completed_tasks_.load(std::memory_order_acquire) == 0) {
-                                threads_done_.release();
+                            if (in_flight_tasks_.load(std::memory_order_acquire) == 0) {
+                                threads_complete_signal_ = true;
```

Comment:
[nit] for consistency you may want to explicitly call store here

Reply:
Good catch! I'll comb over the code a bit more and tidy up things.